elide-dbmanager/elide-dbmanager-hibernate3/src/main/java/com/yahoo/elide/dbmanagers/hibernate3/HibernateTransaction.java

```diff
@@ -0,0 +1,212 @@
+import java.io.IOException;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Hibernate Transaction implementation.
+ */
+public class HibernateTransaction implements DatabaseTransaction {
+    private HibernateManager hibernateManager;
+    private final Transaction transaction;
+    private MultivaluedMap<String, String> queryParams;
+    private final LinkedHashSet<Runnable> deferredTasks = new LinkedHashSet<>();
+
+    /**
+     * Instantiates a new Hibernate transaction.
+     *
+     * @param transaction the transaction
+     */
+    public HibernateTransaction(HibernateManager hibernateManager, Transaction transaction,
+                                MultivaluedMap<String, String> queryParams) {
+        this.hibernateManager = hibernateManager;
+        this.transaction = transaction;
+        this.queryParams = queryParams;
+    }
+
+    public HibernateTransaction(HibernateManager hibernateManager, Transaction transaction) {
+        this.hibernateManager = hibernateManager;
+        this.transaction = transaction;
+        this.queryParams = null;
+    }
+
+    @Override
+    public void delete(Object object) {
+        deferredTasks.add(() -> hibernateManager.getSession().delete(object));
+    }
+
+    @Override
+    public void save(Object object) {
+        deferredTasks.add(() -> hibernateManager.getSession().saveOrUpdate(object));
+    }
+
+    @Override
+    public void flush() {
+        try {
+            deferredTasks.forEach(Runnable::run);
+            deferredTasks.clear();
+            hibernateManager.getSession().flush();
+        } catch (HibernateException e) {
+            throw new TransactionException(e);
+        }
+    }
+
+    @Override
+    public void commit() {
+        try {
+            this.flush();
+            this.transaction.commit();
+        } catch (HibernateException e) {
+            throw new TransactionException(e);
+        }
+    }
+
+    @Override
+    public <T> T createObject(Class<T> entityClass) {
+        try {
+            T object = entityClass.newInstance();
+            deferredTasks.add(() -> hibernateManager.getSession().persist(object));
+            return object;
+        } catch (InstantiationException | IllegalAccessException e) {
+            return null;
+        }
+    }
+
+    @Override
+    public <T> T loadObject(Class<T> loadClass, String id) {
+        @SuppressWarnings("unchecked")
+        T record = (T) hibernateManager.getSession().load(loadClass, Long.valueOf(id));
+        try {
+            Hibernate.initialize(record);
+        } catch (ObjectNotFoundException e) {
+            return null;
+        }
+        return record;
+    }
+
+    @Override
+    public <T> Iterable<T> loadObjects(Class<T> loadClass) {
+        @SuppressWarnings("unchecked")
+        Iterable<T> list = new HibernateManager.ScrollableIterator(
+                applyPredicate(hibernateManager.getSession().createCriteria(loadClass), queryParams)
```

Comment:
I'm not sure this covers the case of collections that are deeper than the root collection.